module/zfs/zfs_fm.c

```diff
@@ -1073,11 +1355,61 @@ zfs_post_state_change(spa_t *spa, vdev_t *vd, uint64_t laststate)
 #endif
 }
 
-#if defined(_KERNEL)
+#ifdef _KERNEL
+void
+zfs_ereport_init(void)
+{
+	mutex_init(&recent_events_lock, NULL, MUTEX_DEFAULT, NULL);
+	list_create(&recent_events_list, sizeof (recent_events_node_t),
+	    offsetof(recent_events_node_t, re_list_link));
+	avl_create(&recent_events_tree,  recent_events_compare,
+	    sizeof (recent_events_node_t), offsetof(recent_events_node_t,
+	    re_tree_link));
+}
+
+/*
+ * This 'early' fini needs to run before zfs_fini() which on Linux waits
+ * for the system_delay_taskq to drain.
+ */
+void
+zfs_ereport_taskq_fini(void)
+{
+	mutex_enter(&recent_events_lock);
+	if (recent_events_cleaner_tqid != 0) {
+		taskq_cancel_id(system_delay_taskq, recent_events_cleaner_tqid);
+		recent_events_cleaner_tqid = 0;
+	}
+	mutex_exit(&recent_events_lock);
+}
+
+void
+zfs_ereport_fini(void)
+{
+	recent_events_node_t *entry;
+
+	if (duplicate_detected_count > 0)
+		cmn_err(CE_NOTE, "zfs_ereport_fini, %d duplicates detected\n",
+		    duplicate_detected_count);
```

Comment:
Why log this to the console as an error when duplicates aren't entirely unexpected?  If this is something we want to track (and I think it is), I'd suggest instead adding it to the `/proc/spl/kstat/zfs/fm` kstat.  The kstat is registered in `fm_init()`, see the `erpt_kstat` structure.

```
struct erpt_kstat {
        kstat_named_t   erpt_dropped;           /* num erpts dropped on post */
        kstat_named_t   erpt_set_failed;        /* num erpt set failures */
        kstat_named_t   fmri_set_failed;        /* num fmri set failures */
        kstat_named_t   payload_set_failed;     /* num payload set failures */
};
```

Reply:
Great idea!  Added it as a kstat:
```name                            type data
erpt-dropped                    4    0
erpt-set-failed                 4    0
fmri-set-failed                 4    0
payload-set-failed              4    0
erpt-duplicates                 4    1512
```